File: helium/src/main/groovy/com/stanfy/helium/handler/codegen/objectivec/ObjCEntitiesGenerator.java
package com.stanfy.helium.handler.codegen.objectivec;

import com.stanfy.helium.handler.Handler;
import com.stanfy.helium.handler.codegen.BaseGenerator;
import com.stanfy.helium.handler.codegen.objectivec.builder.DefaultObjCProjectBuilder;
import com.stanfy.helium.model.Project;

import java.io.File;

/**
 * Created by ptaykalo on 8/25/14.
 */
public class ObjCEntitiesGenerator extends BaseGenerator<ObjcEntitiesOptions> implements Handler {

  private DefaultObjCProjectBuilder projectBuilder;

  public ObjCEntitiesGenerator(final File outputFile, final ObjcEntitiesOptions options) {
    super(outputFile, options);
    this.projectBuilder = new DefaultObjCProjectBuilder();
  }

  @Override
  public void handle(final Project project) {
    ObjCProject objCProject = this.projectBuilder.build(project, getOptions());
    new ObjCProjectGenerator(getOutputDirectory(), objCProject).generate();
  }

}

File: codegen/objc/objc-entities/src/main/kotlin/com/stanfy/helium/handler/codegen/objectivec/entity/ObjCEntitiesGenerator.kt
package com.stanfy.helium.handler.codegen.objectivec.entity;

import com.stanfy.helium.handler.Handler
import com.stanfy.helium.handler.codegen.BaseGenerator
import com.stanfy.helium.handler.codegen.objectivec.entity.builder.*
import com.stanfy.helium.handler.codegen.objectivec.entity.httpclient.urlsession.ObjCHTTPClientGenerator
import com.stanfy.helium.handler.codegen.objectivec.entity.mapper.mantle.ObjCMantleMappingsGenerator
import com.stanfy.helium.handler.codegen.objectivec.entity.mapper.sfobjectmapping.ObjCSFObjectMappingsGenerator
import com.stanfy.helium.handler.codegen.objectivec.entity.typemapping.ObjCTypeMappingRegistry
import com.stanfy.helium.model.Project
import java.io.File

/**
 * Created by ptaykalo on 8/25/14.
 */
class ObjCEntitiesGenerator(outputDirectory: File?, options: ObjCEntitiesOptions?) : BaseGenerator<ObjCEntitiesOptions>(outputDirectory, options), Handler {

  private val objCTypeTransformer = ObjCTypeTransformer()
  private val objCPropertyNameTransformer = ObjCPropertyNameTransformer()

  private val projectBuilder = ObjCDefaultProjectBuilder(objCTypeTransformer, objCPropertyNameTransformer)

  override fun handle(project: Project?) {

    val buildContext = ObjCProjectBuildContext(ObjCTypeMappingRegistry(), ObjCPropertyNameTransformer(), options )
    val builder = ObjCProjectBuilder()
    val objCProject = builder.build(project!!, buildContext)
    val httpClientGenerator = ObjCHTTPClientGenerator(buildContext)
    httpClientGenerator.generate(objCProject, project, options)

//    // Generate mappings
//    val mapper = mapperFromOptions(options)
//    mapper?.generate(objCProject, project, options)
//
//    val fileStructureBuilder = ObjCDefaultFileStructureBuilder()
//    val filesStructure = fileStructureBuilder.build(objCProject.classesTree)
//    ObjCProjectGenerator(outputDirectory, filesStructure).generate()
  }

  private fun mapperFromOptions(options: ObjCEntitiesOptions?): ObjCProjectStructureGenerator? {
    return when (options?.mappingsType) {
      ObjCMappingOption.MANTLE -> ObjCMantleMappingsGenerator()
      ObjCMappingOption.SFMAPPING -> ObjCSFObjectMappingsGenerator()
      else -> null
    }
  }

}
